OSD: Add lightweight ShardedOpWQ statistics for worker threads

blinick opened the pull request:
Add lightweight statistics to the ShardedOpWQ worker threads for monitoring and benchmarking.
Each worker thread will track # times dispatched, # times stalled (sleep due to empty work
queue), and the length of the thread's work queue when it dispatches an op.  These stats are
accessible by the admin socket and will be viewable live with a companion utility in
Ceph/cbt utils.


Uses:
* Stats have been used for evaluation of the simple messenger changes (see PR 5211, graphed data). 
* Also presented in perf meeting 7/15/15.  Also see attached gnuplot data.
* Plan to use in evaluating OSD Worker thread to core affinity with various OSD backends.

Graphs from 7/15/15 perf meeting:
![nvme_qd16_iops_vs_runpct_shards](https://cloud.githubusercontent.com/assets/11448123/8947546/19fe4e54-3551-11e5-9f3b-694450e19b9c.png)
![nvme_qd32_iops_vs_runpct_shards](https://cloud.githubusercontent.com/assets/11448123/8947547/2302ff9a-3551-11e5-8f73-f164f2478d05.png)


Signed-off by: Stephen Blinick <stephen.l.blinick@intel.com>

yuyuyu101 commented on src/osd/OSD.cc:
```diff
@@ -1983,6 +2012,14 @@ void OSD::final_init()
+  r = admin_socket->register_command("dump_shard_stats_and_clear", "dump_shard_stats_and_clear",
+				     asok_hook,
+				     "dump shard worker stats, clearing them after");
+  assert(r == 0);
```
osd_op_queue should more accurate to name instead of "shard"?

yuyuyu101 commented on src/osd/OSD.cc:
```diff
@@ -8184,7 +8223,9 @@ void OSD::ShardedOpWQ::_process(uint32_t thread_index, heartbeat_handle_d *hb )
       return;
     }
   }
+  sdata->ql_accum += sdata->pqueue.length(); 
```
Sadly PrioritizedQueue::length isn't a O(1) implementation. We need to consider to avoid to get length each time or modify PrioritizedQueue::length

blinick commented:
Thanks for the quick review!  You're right, shard isn't very specific.. I'll rename the admin socket command name and the function names as well.  

Good point on the queue length function.  I didn't see any performance degradation in my tests, but in almost all cases the QD was between 1-2 which means it probably didn't take much to calculate.   For that same reason, I'm thinking the usefulness here is in whether there was 1, or >1 item on the queue when an op is dispatched.   So I think the best plan for now would be to convert the length check into an additional empty() check after the dequeue.  That should be fairly lightweight and alllow us to calculate the average length of the queue from 0 <= 2.      I will  do the change, verify it makes sense, and followup with another commit.    Thanks

blinick commented:
Updated to change references as suggested (now using dump_opwq_thread_stats).
Replaced call to length() with a second check of empty() after the dequeue has occurred.  This changes the measurement, which now records whether the queue has nothing (0), just the single item (1), or atleast 1 more (2).  In testing above, generally the queue depth is between 1 and 1.5 for all workloads with reasonable values for the # shards.  Only in the extreme case (i.e. OSD with single shard and low performance) is the queue depth higher.  In the new measurement, the measurements in the range of 1.0->1.5 will be identical, but above that indicates a potentially long queue depth.  See attached graph comparing measurements before and after.  (blue was first PR, orange is the revised code) 

![pr_revised_qd_measurement](https://cloud.githubusercontent.com/assets/11448123/9020867/a1713654-37da-11e5-817b-bc3e2411d423.JPG)

blinick commented:
Regarding the ifdef suggestion:  This code would be intended to be used for benchmarking, comparing different back-ends, as well as different OSD parameters.  So it would have usefulness for users running packages from a repository or otherwise distributed code.  Putting under a compile directive means it's only usable by developers building code.   

I compared a disassemble of the latest code and found only a few instructions added (much less than anything op tracker does for example).   If it helps I can provide the disassembly comparison.

yuyuyu101 commented on src/osd/OSD.cc:
```diff
@@ -8185,6 +8224,8 @@ void OSD::ShardedOpWQ::_process(uint32_t thread_index, heartbeat_handle_d *hb )
     }
   }
   pair<PGRef, PGQueueable> item = sdata->pqueue.dequeue();
+  sdata->ql_accum += sdata->pqueue.empty() ? 1 : 2; 
```
what do you think maintaining the ql_accum by ShardedOpWQ itself instead of getting length from pqueue.

blinick replied:
ql_accum is supposed to be the sum of all queue lengths at dispatch time for this shard.  The measurement on a per-shard basis allows you to calculate the distribution of work across shards.  Sometimes we've suspected that one shard gets more backed up than another.  So that is why it's calculated here by each shard's queue length.      To calculate average queue length this number is divided by the number of dispatches that have happened.   Let me know if this doesn't answer the question.

yuyuyu101 commented:
Sorry, I can't understand the graph fully. Is this mean that the blue one plays better than orange one?

blinick commented:
I ran a set of 100% read tests and measured the average queue length during the test for different OSD settings (# shards, # workers per shard) to show the range of values this measurement gets. 

The blue line is how the measurement looked when I was usuing pqueue.length() original code.  The orange line is how the measurements look with the latest commit..  The measurement is less precise because I can't record the entire length of the queue, but it should be performant (as you pointed out pqueue.length may not be O(1).)  

What I wanted to show was that the measurement is still useful this way because anything over 1.5 seems to show the shard is rather backed up.  Snd there is still the seperate measurement of run/sleep % for each shard.   Thanks.

liewegas commented:
This doesn't really worry me as far as performance goes, these are all tiny fixed-cost operations.

This could use perfcounter, but it'd be more code and slower, so this reporting is probably okay.

@athanatos what do you think?

liewegas commented:
Okay, consnensus is we should use perf counters.  One per shard.  This let syou drop teh ad hoc dump functions.

blinick commented:
Thanks for review; I'll update this PR with the changes.

liewegas commented:
pls reopen when updated, thanks!
